_posts/2021-01-05-git-rebase-merge.md

```diff
@@ -0,0 +1,202 @@
+---
+
+title: "Git Rebase Vs Git Merge: Which One Is Better?"
+categories: [craft]
+modified: 2021-01-05
+excerpt: "If you are new to Git and/or GitHub, it's easy to get overwhelmed by the different workflow models you can use. 
+          This post explains the basic fork and pull workflow model that is used on a lot of GitHub repositories." 
+image:
+  auto: 0050-git
+---
+
+In this article we are going to discuss some of the very important commands from [git](https://git-scm.com/) and [GitHub](https://github.com), and how they make the life of developers easy working individually or working in a team, and when to utilize each of them. We will be comparing Git Rebase and Git Merge, and further, we will explore the potential opportunities to incorporate the basic workflow of Git. We will start from the very basic, and then will keep on leveling up.
+
+If you are a beginner to git or GitHub and are looking to understand some of the basic commands used with git like fork, pull, checkout and commit then you should definitely give this [amazing article](https://reflectoring.io/github-fork-and-pull/) a read.
+
+
+
+- Introduction to Git
+- What is Git Rebase?
+- Working of Git Rebase
+- What is Git Merge?
+- Working of Git Merge
+- Git Rebase vs Git Merge
+- Merging Pros and Cons
+- Choosing the Right Method
+- Common Myths
+- Summary
+
+So bear with me. I hope this will be fun and a learning experience for all those getting their hands dirty with Git 😊
+
+## Introduction to Git
+
+An **Open-Source Distributed Version Control System** is known as Git. It can be broken down into the following major components,
+
+**Control System:** It acts as a content tracker for Git. So, basically, Git can be used to store content – it is usually used to store the code, but the other content can also be stored.
+
+**Version Control System:** Since the developers code in parallel so the version control system helps in maintaining a history of changes by providing the features like branches and merges.
+
+**Distributed Version Control System:** The code in the case of Git is present in two types of repositories – the local, and the central. Managing and synchronizing the local and central repositories comes under the shelter of a Distributed Version Control System.
+
+## What is Git Rebase?
+
+Let's have a look at the concept of Git Rebase. It is the way of migrating or combining a sequence of commits to a new base commit. If one considers it in the context of feature branching workflow, then it is more useful and easily visualized. The figure given below describes the general process of Git Rebasing.
+
+![Git Rebasing](/assets/img/posts/git-rebase-merge/git-rebasing-basic.png)

```

Comment:
I have used the graphics from various sources mostly from Youtube videos or various articles, let me know how I can reference these links with the images for copyright reasons.
I have added an explanation for the feature box in the text below the diagram, good suggestion from the user point of view.

Reply:
We could link to the sources for all the graphics, but that bloats the article and is tedious for the reader. 

Also, with the graphics coming from different sources, they look different (I see at least 4 different "styles" of graphics), and they use different concepts (one uses a feature branch, another says "his branch", another uses "server" and "client" branches, ... . That fragments the article and the reader's attention. We want to write an article that makes explaining Merging and Rebasing as simple as we can, and that means that the graphics should use the same style and concepts.

I think the best way would be to re-create the graphics as we need them. This way they can all look alike and we can even align them in wording (i.e. use `main` instead of `master` everywhere, and use a branch named `feature` everywhere, and perhaps use `feature1` and `feature2` instead of the `client` and `server` branches, because the client/server example is very specific and many readers might not relate to that).

I have given it a try with the free editor [yEd](https://www.yworks.com/products/yed) and it's rather easy to create those graphs with it. I attached an example to this comment that you can open with yed. Please use yEd to re-create the graphics. Please push the source files (*.graphml) to the repository as well, so I can do minor adjustments myself and don't have to get back to you for every change I might want to make.

Sorry to put this on you, but I want to make the article the best version it can be.

[git.graphml.zip](https://github.com/reflectoring/reflectoring.github.io/files/5791793/git.graphml.zip)

Reply:
@thombergs I have created new images for each of the images used with consistent naming and coloring however there are alignment issues for some of the images and I hope you can sort them out.
This article has already taken a lot more of my time especially with the images than I had thought initially because this is the first time I am creating artwork while creating content.
I hope you will consider these during the compensation part.
Thanks for the suggestions and help.